File: dashboard/replay.js
/* Decision replay engine -- a line-for-line port of src/decision/hybrid.py
 * (EscalationMonitor) and src/decision/replay.py (hybrid_lower_bound,
 * unit_outcome, age_outcome, aggregate). Shared by the dashboard and by
 * scripts/check_dashboard.mjs, which verifies it against the Python results.
 */
(function (root) {
  "use strict";

  const POLICIES = ["hybrid", "edge_conformal", "edge_point", "twin_always", "twin_point",
                    "lstm_mc", "lstm_point", "oracle"];

  function decode(deltas, scale) {
    const out = new Float64Array(deltas.length);
    let acc = 0;
    for (let i = 0; i < deltas.length; i++) { acc += deltas[i]; out[i] = acc / scale; }
    return out;
  }

  /* Decode one exported engine into plain Float64Arrays (cached on the object). */
  function engineSeries(e, scale) {
    if (e._s) return e._s;
    const r = scale.rul;
    const s = {
      edge: decode(e.edge, r), elo: decode(e.elo, r), ehi: decode(e.ehi, r),
      t05: decode(e.t05, r), t50: decode(e.t50, r), t95: decode(e.t95, r),
      lstm: decode(e.lstm, r), llo: decode(e.llo, r),
      hi: decode(e.hi, scale.hi),
      sens: e.sens.map((d) => decode(d, scale.sens)),
      n: e.edge.length,
    };
    s.rulTrue = new Float64Array(s.n);
    for (let i = 0; i < s.n; i++) s.rulTrue[i] = e.life - (i + 1);
    e._s = s;
    return s;
  }

  /* EscalationMonitor.step, stateful across calls. */
  function makeMonitor(esc) {
    let last = null;
    let since = esc.resync;
    return function step(edgeRul) {
      since += 1;
      const reasons = [];
      if (esc.lowRul && edgeRul <= esc.watchRul) reasons.push("low_rul");
      if (esc.sharpDrop && last !== null && last - edgeRul >= esc.deltaDrop) reasons.push("sharp_drop");
      if (esc.periodic && since >= esc.resync) reasons.push("periodic_resync");
      if (last === null && reasons.length === 0) reasons.push("initial_sync");
      if (reasons.length) since = 0;
      last = edgeRul;
      return reasons;
    };
  }

  /* Per-cycle lower bound, twin-synced flags and trigger reasons for one policy. */
  function lowerBound(s, policy, cfg) {
    const n = s.n;
    const synced = new Uint8Array(n);
    let lo;
    let reasons = null;
    switch (policy) {
      case "edge_point": lo = s.edge; break;
      case "edge_conformal": lo = s.elo; break;
      case "lstm_point": lo = s.lstm; break;
      case "lstm_mc": lo = s.llo; break;
      case "twin_point": lo = s.t50; synced.fill(1); break;
      case "twin_always": lo = s.t05; synced.fill(1); break;
      case "oracle": lo = s.rulTrue; break;
      case "hybrid": {
        const step = makeMonitor(cfg);
        lo = Float64Array.from(s.elo);
        reasons = new Array(n);
        for (let i = 0; i < n; i++) {
          const r = step(s.edge[i]);
          reasons[i] = r;
          if (r.length) {
            synced[i] = 1;
            lo[i] = cfg.fusion === "twin" ? s.t05[i] : Math.min(s.t05[i], s.elo[i]);
          }
        }
        break;
      }
      default: throw new Error("unknown policy " + policy);
    }
    return { lo, synced, reasons };
  }

  function zoneOf(lo, cfg) {
    if (lo <= cfg.ground) return "GROUND_NOW";
    if (lo <= cfg.watch) return "SCHEDULE_MAINTENANCE";
    if (lo <= cfg.safe) return "WATCH";
    return "SAFE";
  }

  function unitOutcome(life, lo, synced, ground) {
    const n = lo.length;
    let stop = n - 1;
    for (let i = 0; i < n; i++) if (lo[i] <= ground) { stop = i; break; }
    const rulAtStop = life - (stop + 1);
    const failed = rulAtStop <= 0;
    let syncs = 0;
    for (let i = 0; i <= stop; i++) syncs += synced[i];
    return { life, operated: failed ? life : stop + 1, failed, wasted: failed ? 0 : rulAtStop,
             syncs, cycles: stop + 1, stop };
  }

  function ageOutcome(life, age) {
    const failed = life <= age;
    return { life, operated: failed ? life : age, failed, wasted: failed ? 0 : life - age,
             syncs: 0, cycles: Math.min(life, age), stop: Math.min(life, age) - 1 };
  }

  function aggregate(outcomes, costs) {
    let fail = 0, wasted = 0, operated = 0, life = 0, syncs = 0, cycles = 0;
    for (const o of outcomes) {
      if (o.failed) fail++; else wasted += o.wasted;
      operated += o.operated; life += o.life; syncs += o.syncs; cycles += o.cycles;
    }
    const pm = outcomes.length - fail;
    return {
      n_units: outcomes.length, failures: fail, failure_rate: fail / outcomes.length,
      mean_wasted_life: pm ? wasted / pm : 0, life_utilization: operated / life,
      cost_rate_x1000: 1000 * (costs.failure * fail + costs.preventive * pm) / operated,
      twin_sync_rate: syncs / cycles,
    };
  }

  function evaluatePolicy(ds, policy, cfg, scale) {
    const outcomes = [];
    for (const e of ds.engines) {
      if (policy === "age_replacement") { outcomes.push(ageOutcome(e.life, e.age)); continue; }
      if (policy === "run_to_failure") { outcomes.push(ageOutcome(e.life, 1e9)); continue; }
      const s = engineSeries(e, scale);
      const { lo, synced } = lowerBound(s, policy, cfg);
      outcomes.push(unitOutcome(e.life, lo, synced, cfg.ground));
    }
    return { summary: aggregate(outcomes, { preventive: cfg.costPreventive, failure: cfg.costFailure }), outcomes };
  }

  function defaultConfig(defaults) {
    return {
      safe: defaults.safe, watch: defaults.watch, ground: defaults.ground,
      watchRul: defaults.watchRul, deltaDrop: defaults.deltaDrop, resync: defaults.resync,
      lowRul: true, sharpDrop: true, periodic: true, fusion: "min",
      costPreventive: defaults.costPreventive, costFailure: defaults.costFailure,
    };
  }

  const api = { POLICIES, decode, engineSeries, makeMonitor, lowerBound, zoneOf, unitOutcome,
                ageOutcome, aggregate, evaluatePolicy, defaultConfig };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else root.TwinReplay = api;
})(typeof window !== "undefined" ? window : globalThis);

File: src/decision/replay.py
"""Replay maintenance policies over run-to-failure traces and score them.

Outcome model (per engine):
  - The first cycle whose action is GROUND_NOW removes the engine for
    preventive maintenance (cost C_p). Life left on the table = true RUL then.
  - If the engine reaches its failure cycle (true RUL = 0) first, it is an
    unplanned failure (cost C_f).
Fleet metric is the long-run cost rate from renewal-reward theory,
    cost_rate = total cost / total cycles operated,
the same objective that defines classical age-replacement (Barlow & Hunter, 1960).
"""

from dataclasses import dataclass

import numpy as np
import pandas as pd

from src.decision.hybrid import EscalationMonitor, EscalationThresholds
from src.decision.policy import DecisionThresholds

Z90 = 1.6449  # one-sided 95% normal quantile -> matches the 5%-quantile lower bounds


@dataclass(frozen=True)
class Costs:
    preventive: float = 1.0
    failure: float = 10.0


def hybrid_lower_bound(g: pd.DataFrame, escalation: EscalationThresholds, fusion: str = "min",
                       twin_lo: str = "twin_q05") -> tuple[np.ndarray, np.ndarray, list[list[str]]]:
    """Replay the escalation monitor over one engine's trace.

    Between syncs the edge conformal lower bound is used. At a sync:
      fusion="twin": trust the twin's posterior lower quantile alone
      fusion="min":  take the more cautious of the two calibrated bounds; each is
                     a one-sided 95% bound, so the minimum keeps >= 90% coverage
                     by the union bound while inheriting the edge model's
                     sharpness near failure and the twin's long-horizon tracking.
    """
    monitor = EscalationMonitor(escalation)
    edge, edge_lo, tlo = g["edge"].to_numpy(), g["edge_lo"].to_numpy(), g[twin_lo].to_numpy()
    lo = edge_lo.astype(float).copy()
    synced = np.zeros(len(g), dtype=bool)
    reasons_all = []
    for i in range(len(g)):
        reasons = monitor.step(float(edge[i]))
        reasons_all.append(reasons)
        if reasons:
            synced[i] = True
            lo[i] = tlo[i] if fusion == "twin" else min(tlo[i], edge_lo[i])
    return lo, synced, reasons_all


def lower_bound(g: pd.DataFrame, policy: str, **kw) -> tuple[np.ndarray, np.ndarray]:
    """Returns (lower bound per cycle, twin-synced flag per cycle)."""
    none = np.zeros(len(g), dtype=bool)
    if policy == "edge_point":
        return g["edge"].to_numpy(), none
    if policy == "edge_conformal":
        return g["edge_lo"].to_numpy(), none
    if policy == "lstm_point":
        return g["lstm"].to_numpy(), none
    if policy == "lstm_mc":
        return (g["lstm_mc_mean"] - Z90 * g["lstm_mc_std"]).to_numpy(), none
    if policy == "twin_point":
        return g["twin_q50"].to_numpy(), ~none
    if policy == "twin_always":
        return g["twin_q05"].to_numpy(), ~none
    if policy == "oracle":
        return g["rul_true"].to_numpy(), none
    if policy.startswith("hybrid"):
        lo, synced, _ = hybrid_lower_bound(g, kw.get("escalation", EscalationThresholds()),
                                           kw.get("fusion", "min"), kw.get("twin_lo", "twin_q05"))
        return lo, synced
    raise ValueError(policy)


def unit_outcome(g: pd.DataFrame, lo: np.ndarray, synced: np.ndarray, ground: float) -> dict:
    life = int(g["cycle"].iloc[-1])
    hits = np.flatnonzero(lo <= ground)
    stop = int(hits[0]) if len(hits) else len(g) - 1
    rul_at_stop = float(g["rul_true"].iloc[stop])
    failed = rul_at_stop <= 0
    operated = life if failed else int(g["cycle"].iloc[stop])
    return {"unit": int(g["unit"].iloc[0]), "life": life, "operated": operated, "failed": bool(failed),
            "wasted": 0.0 if failed else rul_at_stop, "syncs": int(synced[: stop + 1].sum()),
            "cycles": stop + 1}


def age_outcome(life: int, age: int) -> dict:
    failed = life <= age
    return {"life": life, "operated": life if failed else age, "failed": bool(failed),
            "wasted": 0.0 if failed else float(life - age), "syncs": 0, "cycles": min(life, age)}


def optimal_age(lifetimes: np.ndarray, costs: Costs = Costs()) -> int:
    """Barlow-Hunter age replacement on the empirical lifetime distribution."""
    lifetimes = np.asarray(lifetimes)
    best_age, best_rate = None, np.inf
    for a in range(1, int(lifetimes.max()) + 1):
        failed = lifetimes <= a
        cost = costs.failure * failed.sum() + costs.preventive * (~failed).sum()
        rate = cost / np.minimum(lifetimes, a).sum()
        if rate < best_rate:
            best_age, best_rate = a, rate
    return best_age


def aggregate(outcomes: list[dict], costs: Costs = Costs()) -> dict:
    df = pd.DataFrame(outcomes)
    n_fail = int(df["failed"].sum())
    n_pm = len(df) - n_fail
    total_cost = costs.failure * n_fail + costs.preventive * n_pm
    return {
        "n_units": len(df),
        "failures": n_fail,
        "failure_rate": n_fail / len(df),
        "mean_wasted_life": float(df.loc[~df["failed"], "wasted"].mean()) if n_pm else 0.0,
        "life_utilization": float(df["operated"].sum() / df["life"].sum()),
        "cost_rate_x1000": 1000 * total_cost / float(df["operated"].sum()),
        "twin_sync_rate": float(df["syncs"].sum() / df["cycles"].sum()),
    }


def evaluate_policy(traces: pd.DataFrame, policy: str, thresholds: DecisionThresholds = DecisionThresholds(),
                    costs: Costs = Costs(), **kw) -> tuple[dict, list[dict]]:
    outcomes = []
    for _, g in traces.groupby("unit", sort=True):
        lo, synced = lower_bound(g, policy, **kw)
        outcomes.append(unit_outcome(g, lo, synced, thresholds.ground))
    return aggregate(outcomes, costs), outcomes


def evaluate_age_policy(traces: pd.DataFrame, ages_by_fold: dict[int, int], costs: Costs = Costs()) -> dict:
    lives = traces.groupby("unit").agg(life=("cycle", "max"), fold=("fold", "first"))
    outcomes = [age_outcome(int(r.life), ages_by_fold[int(r.fold)]) for r in lives.itertuples()]
    return aggregate(outcomes, costs)
